### DISCUSSION TITLE
Handling Multiple Schemas

### DISCUSSION THREAD

**Discussion by hxk1633** [diesel-rs/diesel#4359] [Q&A]
I'm not sure if this "sub schema" feature released in version 2.2.0 is the right tool for what I'm doing. I have created a new schema in my migration files called "v2" and I have an old default schema (public) that I need to query with diesel.  These are the diesel.toml and schema files.

```
# For documentation on how to configure this file,
# see https://diesel.rs/guides/configuring-diesel-cli

[print_schema]
file = "src/diesel_schema/public.rs"
custom_type_derives = ["diesel::query_builder::QueryId"]

[print_schema.v2]
file = "src/diesel_schema/v2.rs"
custom_type_derives = ["diesel::query_builder::QueryId"]

[migrations_directory]
dir = "migrations"
```

`src/diesel_schema/v2.rs`
```
// @generated automatically by Diesel CLI.

pub mod v2 {
    pub mod sql_types {
        #[derive(diesel::query_builder::QueryId, diesel::sql_types::SqlType)]
        #[diesel(postgres_type(name = "device_status", schema = "v2"))]
        pub struct DeviceStatus;
    }

    diesel::table! {
        v2.account (id) {
           ....
        }
    }

    diesel::table! {
        v2.contact (id) {
         ...
        }
    }

    diesel::table! {
        v2.device (id) {
             ....
        }
    }

    diesel::joinable!(device -> account (account_id));

    diesel::allow_tables_to_appear_in_same_query!(
        account,
        contact,
        device
    );
}
```

`src/diesel_schema/public.rs`
```
pub mod public {
    pub mod sql_types {
        #[derive(diesel::query_builder::QueryId, diesel::sql_types::SqlType)]
        #[diesel(postgres_type(name = "device_status", schema = "public"))]
        pub struct DeviceStatus;
    }

    diesel::table! {
        public.account (id) {
           ....
        }
    }

    diesel::table! {
        public.contact (id) {
         ...
        }
    }

    diesel::table! {
        public.device (id) {
             ....
        }
    }

    diesel::joinable!(device -> account (account_id));

    diesel::allow_tables_to_appear_in_same_query!(
        account,
        contact,
        device
    );
}
```

I'm not sure if the schema names "public" and "v2" have to be included in the schema files. Is the schema name after "print_schema" in diesel.toml enough for me to start querying from both public and v2 schema in my code?

**Comment by forest1102** [comment]
If no key is provided, it is treated as sub print schema with  `"default"` key name.  
So, in this case, there are two print_schema, default and v2.

**Reply by hxk1633** [reply]
Oh ok, thanks. But with this feature, I'll be able to query both schemas in the same database at the same time?

**Reply by hxk1633** [reply]
I just want to make sure I understand what a sub schema is and what it's used for. I assume these sub schemas map to schemas we define in our database?

**Reply by forest1102** [reply]
Main purpose is by splitting schema.rs into different crates, to reduce compilation time and limit what database to use in repository layer.

**Reply by weiznich** [reply]
> I just want to make sure I understand what a sub schema is and what it's used for. I assume these sub schemas map to schemas we define in our database?

They are just arbitrary named identifiers for diesel to identify what to write into a `schema.rs` file at a given path. They can contain all tables from a single database schema or they can contain additional table include/exclude filters to strip down what's included. Also several sub-schemas can refer to the same database schema.
